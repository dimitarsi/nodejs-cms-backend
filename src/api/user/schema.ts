import { JSONSchemaType } from "ajv";
import { User } from "~/models/user";
import ajv from "~/schema/core";

const notActive = {
  $or: [
    { isActive: { $exists: false } },
    { isActive: false },
    { isActive: null },
  ],
  // isActive: false
};

export const validateIsInactive = ajv.compile<JSONSchemaType<{ id: string }>>({
  type: "object",
  properties: {
    id: {
      type: "string",
      found: {
        in: "users",
        extra: notActive,
      },
    },
  },
  $async: true,
  required: ["id"],
});

export const validateCreateUser = ajv.compile<JSONSchemaType<User>>({
  $async: true,
  type: "object",
  properties: {
    firstName: {
      type: "string"
    },
    lastName: {
      type: "string",
    },
    password: {
      type: "string",
      minLength: 8
    },
    email: {
      type: "string",
      format: "email",
      found: {in: "users:email", notIn: true}
    },
    gender: {
      enum: ["f", "m", "other"]
    }
  },
  required: ["firstName", "lastName", "password", "email"],
  additionalProperties: false
})
