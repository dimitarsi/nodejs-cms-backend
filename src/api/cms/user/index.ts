import { FastifyInstance } from "fastify"
import { getAllUsers } from "./actions/getAll"
import createUser from "./actions/createUsers"
import deleteUser from "./actions/deleteUser"
import updateUser from "./actions/updateUser"
import auth from "@middleware/auth"
import activateUser from "./actions/activateUser"

export default function (
  instance: FastifyInstance,
  options: never,
  done: Function
) {
  activateUser(instance)

  instance.register((instance, options, done) => {
    auth(instance, { isAdmin: true }, () => {})

    getAllUsers(instance)

    createUser(instance)

    deleteUser(instance)

    updateUser(instance)

    done()
  })

  done()
}
