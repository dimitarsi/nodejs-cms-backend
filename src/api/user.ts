import controller from "./controller"
import bcrypt from "bcrypt"
import {rounds} from "../config/config"

export interface User {
  firstName: string
  lastName: string
  email: string
  isActive: boolean
  isAdmin: boolean
  password: string
}

export default controller<User>('users', (crud, _collection) => ({
  ...crud,
  create: (data: Partial<User>) => {
    if(!data.password) {
      throw "Invalid password"
    }

    return crud.create({
      ...data,
      password: bcrypt.hashSync(data.password, rounds),
    });
  },
  update: (id: string | number, data: Partial<User>) => {
    if(!data.password) {
      return crud.update(id, data)
    }

    return crud.update(id, {
      ...data,
      password: bcrypt.hashSync(data.password, rounds),
    });
  }
}))