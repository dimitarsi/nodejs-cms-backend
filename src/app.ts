import "./config"
import cmsAPI from "~/api/cms"
import "~/middleware/auth"
import fastify from "fastify"
import repo from "~/repo"
import publicApiAuth from "~/api/public"
import fastifyPlugin from "fastify-plugin"
import cmsApiSchema from "~/schema/cms-api"

const app = fastify({
  logger: {
    level: "info",
    file: "./server.log",
  },
})

app.register(fastifyPlugin(cmsApiSchema))

app.register(fastifyPlugin(repo), {
  dbName: process.env.DB_NAME || "plenty_cms",
})

app.get("/health", (_r, reply) => {
  reply.code(200).send({})
})

app.register(cmsAPI)
app.register(publicApiAuth)

export default app
