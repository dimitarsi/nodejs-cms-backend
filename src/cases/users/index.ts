import { ensureObjectId } from "~/helpers/objectid"
import { v4 } from "uuid"
import { ObjectId } from "mongodb"
import type {
  UsersRepo,
  AccessTokenRepo,
  ProjectsRepo,
  InvitationsRepo,
} from "~/repo"

function createUserCaseFrom(
  users: UsersRepo,
  accessToken: AccessTokenRepo,
  projects: ProjectsRepo,
  invitations: InvitationsRepo
) {
  const getProjectsFromToken = async (tokenHeader: string) => {
    const token = await accessToken.findToken(tokenHeader)

    if (!token || !token.userId) {
      // reply.code(404).send({ message: "Not Found or token has expired" })
      // return
      return []
    }

    const user = await users.getById(token.userId)

    if (!user?.projects || !user.projects.length) {
      // reply.code(404).send({
      //   message: "Not Found or User is not assigned to any projects",
      // })
      return []
    }

    return await projects.getAllByIds(user.projects)
  }

  const acceptInvitation = async (
    tokenHeader: string,
    invitationToken: string
  ) => {
    const token = await accessToken.findToken(tokenHeader)

    if (!token || !token.userId) {
      // reply.code(404).send({ message: "Not Found or token has expired" })
      // return
      return []
    }

    const user = await users.getById(token?._id)

    const invitation = await invitations.getCollection().findOne({
      invites: {
        $elemMatch: {
          token: invitationToken,
          userEmail: user?.email || "n/a",
          expires: { $gt: new Date() },
        },
      },
    })

    if (!invitation) {
      return null
    }

    // assing user to the project
    await users.getCollection().updateOne(
      {
        _id: new ObjectId(token.userId),
      },
      {
        $push: {
          projects: ensureObjectId(invitation?.projectId),
        },
      }
    )

    // remove the invitaiton
    await invitations.deleteInvitation(invitation._id)
  }

  return {
    getProjectsFromToken,
    acceptInvitation,
  }
}

export default createUserCaseFrom
