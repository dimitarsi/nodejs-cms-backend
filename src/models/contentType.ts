import { ObjectId } from "mongodb"
import slugify from "~/helpers/slugify"

export interface ContentType {
  _id?: ObjectId
  name: string
  slug: string
  type: "root" | "composite" | string
  children?: ContentType[] | null
  /**
   * Flag the ContentType as freezed, meaning should not be allowed to change.
   * This is enforced only on the client, as a measure when referencing other "root"
   * ContentTypes.
   */
  freezed?: boolean
  originalName?: string
  originalSlug?: string
}

export const textContentType = (name: string): ContentType => ({
  name,
  slug: slugify(name),
  type: "text",
  children: null
})

export const freezeAllChildren = (contentType: Pick<ContentType, 'children' | 'freezed'>) => {
  contentType.freezed = true;
  contentType.children?.forEach(c => freezeAllChildren(c))
}

export const unfeezeAllChildren = (contentType: ContentType) => {
  contentType.freezed = true
  contentType.children?.forEach((c) => unfeezeAllChildren(c))
}

/**
 * Helper method to easily compose contentTypes for tests and seeding
 */
export const compositeContentType = (name: string, isRoot = false) => {

  const rootContentType: ContentType = {
    name,
    slug: slugify(name),
    type: isRoot ? "root" : "composite",
    children: [],
    freezed: false
  }

  const chainable = {
    __rootContentType: rootContentType,
    add(contentType: ContentType) {

      if (rootContentType.freezed) {
        contentType.freezed = true

      }

      rootContentType.children!.push(contentType)
      return chainable
    },
    clone() {
      const cloned = compositeContentType(name, isRoot)
      // TODO: add a proper cloning
      cloned.__rootContentType.children = JSON.parse(JSON.stringify(rootContentType.children)) 

      return cloned;
    },
    /**
     * convenience method when setuping up tests
     */
    clearChildren() {
      rootContentType.children = []
    },
    getType() {
      return rootContentType
    },
    freeze() {
      freezeAllChildren(rootContentType);
      return chainable
    },
    unfreeze() {
      freezeAllChildren(rootContentType)
      return chainable
    }
  }

  return chainable;
}