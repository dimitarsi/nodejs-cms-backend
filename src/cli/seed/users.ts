import { validateCreateUser } from "@api/cms/user/schema"
import userRepo from "~/repo/users"
import ajv from "~/schema/core"

const validateEmailIsUnique = ajv.compile({
  type: "object",
  $async: true,
  properties: {
    email: {
      type: "string",
      format: "email",
      found: {in:"users:email", notIn: true}
    }
  }
})

export default async function seedUsers() {
  const data = [{
    firstName: 'Admin',
    lastName: 'Root',
    isAdmin: true,
    isActive: true,
    email: "admin@gmail.com",
    password: "1234567890"
  }]

  for(const userData of data) {
    const valid = await validateEmailIsUnique(userData)

    if (valid) {
      await userRepo.create(userData)    
    } else {
      console.error(validateCreateUser.errors)
    }
  }

}