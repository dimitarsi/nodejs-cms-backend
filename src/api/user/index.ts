import { withDefaultRoutes } from "../controller";
import users from "../../repo/users";
import express from "express";
import validate from "./schema";

const app = express();

withDefaultRoutes(app, users);

app.post("/:id/activate", async (req, res) => {
  try {
    const valid = await validate(req.params);
    
    if (!valid) {
      res.statusCode = 405;
      res.json(validate.errors);
      return;
    }
  } catch (er) {
    res.statusCode = 500;
    res.json({
      error: er
    });
    return;
  }

  await users.activate(req.params.id);

  res.json({ success: true });
});

export default app;
