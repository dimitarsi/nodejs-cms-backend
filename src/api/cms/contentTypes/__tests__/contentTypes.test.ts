import { MongoClient, type WithId } from "mongodb"
import { compositeContentType } from "./../../../../models/contentType"
import { describe, test, expect } from "@jest/globals"
import supertest from "supertest"
import app from "../../../../app"
import users from "~/repo/users"
import accessTokens from "~/repo/accessTokens"
import seedUsers from "~/cli/seed/users"
import { User } from "~/models/user"

describe("ContentTypes", () => {
  const mongoClient = new MongoClient(
    process.env.MONGO_URL || "mongodb://root:example@localhost:27017"
  )
  const db = mongoClient.db(process.env.DB_NAME)

  afterAll(() => {
    app.close()
    mongoClient.close()
  })

  afterEach(async () => {
    await db.collection("contentTypes").deleteMany({})
  })

  describe("Needs Authentication", () => {
    test("GET /content-types", async () => {
      await app.ready()

      await supertest(app.server).get("/content-types").expect(403)
    })

    test("GET /content-types/:id", async () => {
      await app.ready()

      await supertest(app.server).get("/content-types/foobar").expect(403)
    })

    test("POST /content-types", async () => {
      await app.ready()

      await supertest(app.server)
        .post("/content-types")
        .send({
          name: "foo",
          slug: "foo",
          type: "root",
          repeated: true,
          children: [],
        })
        .expect(403)
    })

    test("PATCH /content-types/foobar", async () => {
      await app.ready()

      await supertest(app.server)
        .patch("/content-types/foobar")
        .send({
          hello: "world",
        })
        .expect(403)
    })

    test("DELETE /content-types/foobar", async () => {
      await app.ready()

      await supertest(app.server).delete("/content-types/foobar").expect(403)
    })
  })

  describe("Only Admin users can create, update, delete and inspect content types", () => {
    const usersRepo = users(db)
    const accessTokensRepo = accessTokens(db)

    beforeEach(async () => {
      await usersRepo.deleteAll()
      await accessTokensRepo.deleteAll()

      await seedUsers(db)
      const user = (await usersRepo
        .getCollection()
        .findOne({ isAdmin: true })) as WithId<User>

      await accessTokensRepo.findOrCreateAccessToken(
        user._id,
        { isAdmin: true },
        process.env.TEST_ADMIN_ACCESS_TOKEN
      )
    })

    describe("As Admin user", () => {
      test("GET /content-types", async () => {
        await app.ready()

        supertest(app.server)
          .get("/content-types")
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(200)
      })

      test("POST /content-types", async () => {
        await app.ready()

        const resp = await supertest(app.server)
          .post("/content-types")
          .send(compositeContentType("test01").getType())
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(201)

        expect(resp.body).toBeDefined()
        expect(resp.body).toHaveProperty("_id")
        expect(resp.body).toHaveProperty("type")
        expect(resp.body).toHaveProperty("name")
        expect(resp.body).toHaveProperty("slug")
        expect(resp.body).toHaveProperty("children")
        expect(resp.body).toHaveProperty("repeated")

        const { _id, ...body } = resp.body
        expect(body).toMatchSnapshot()

        expect(resp.headers["location"]).toMatch(/^\/content-types\/(.+)$/)
      })

      test("GET /content-types/:idOrSlug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post("/content-types")
          .send(compositeContentType("test01").getType())
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(201)

        expect(resp.body).toBeDefined()

        const resultFromSlug = await supertest(app.server)
          .get(`/content-types/${resp.body.slug}`)
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(200)

        const resultFromId = await supertest(app.server)
          .get(`/content-types/${resp.body._id}`)
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(200)

        expect(resultFromId.body._id).toBeDefined()
        expect(resultFromSlug.body._id).toBeDefined()
        expect(resultFromId.body._id).toBe(resultFromSlug.body._id)
      })

      test("PATCH /content-types/:idOrSlug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post("/content-types")
          .send(compositeContentType("test01").getType())
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .patch(`/content-types/${resp.body.slug}`)
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .send({
            name: "Hello World",
          })
          .expect(200)

        await supertest(app.server)
          .patch(`/content-types/${resp.body._id}`)
          .send({
            name: "Hello World 2",
          })
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(200)
      })

      test("DELETE /content-types/:id", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post("/content-types")
          .send(compositeContentType("test01").getType())
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .delete(`/content-types/${resp.body._id}`)
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(200)
      })

      test("DELETE /content-types/:slug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post("/content-types")
          .send(compositeContentType("test01").getType())
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .delete(`/content-types/${resp.body.slug}`)
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(200)
      })
    })

    describe("As non-Admin user", () => {
      test("GET /content-types", async () => {
        await app.ready()

        supertest(app.server)
          .get("/content-types")
          .set("X-Access-Token", process.env.TEST_NON_ADMIN_ACCESS_TOKEN!)
          .expect(403)
      })

      test("POST /content-types", async () => {
        await app.ready()

        const resp = await supertest(app.server)
          .post("/content-types")
          .send(compositeContentType("test01").getType())
          .set("X-Access-Token", process.env.TEST_NON_ADMIN_ACCESS_TOKEN!)
          .expect(403)
      })

      test("GET /content-types/:idOrSlug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post("/content-types")
          .send(compositeContentType("test01").getType())
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .get(`/content-types/${resp.body.slug}`)
          .set("X-Access-Token", process.env.TEST_NON_ADMIN_ACCESS_TOKEN!)
          .expect(403)

        await supertest(app.server)
          .get(`/content-types/${resp.body._id}`)
          .set("X-Access-Token", process.env.TEST_NON_ADMIN_ACCESS_TOKEN!)
          .expect(403)
      })

      test("PATCH /content-types/:idOrSlug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post("/content-types")
          .send(compositeContentType("test01").getType())
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .patch(`/content-types/${resp.body.slug}`)
          .set("X-Access-Token", process.env.TEST_NON_ADMIN_ACCESS_TOKEN!)
          .send({
            name: "Hello World",
          })
          .expect(403)

        await supertest(app.server)
          .patch(`/content-types/${resp.body._id}`)
          .send({
            name: "Hello World 2",
          })
          .set("X-Access-Token", process.env.TEST_NON_ADMIN_ACCESS_TOKEN!)
          .expect(403)
      })

      test("DELETE /content-types/:id", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post("/content-types")
          .send(compositeContentType("test01").getType())
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .delete(`/content-types/${resp.body._id}`)
          .set("X-Access-Token", process.env.TEST_NON_ADMIN_ACCESS_TOKEN!)
          .expect(403)
      })

      test("DELETE /content-types/:slug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post("/content-types")
          .send(compositeContentType("test01").getType())
          .set("X-Access-Token", process.env.TEST_ADMIN_ACCESS_TOKEN!)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .delete(`/content-types/${resp.body.slug}`)
          .set("X-Access-Token", process.env.TEST_NON_ADMIN_ACCESS_TOKEN!)
          .expect(403)
      })
    })
  })
})
