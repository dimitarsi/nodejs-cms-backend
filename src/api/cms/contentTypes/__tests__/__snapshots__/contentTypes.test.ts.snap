// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContentTypes > Only Admin users can create, update, delete and inspect content types > As Admin user > POST /content-types 1`] = `
{
  "children": [],
  "freezed": false,
  "name": "test01",
  "repeated": false,
  "slug": "test01",
  "type": "root",
}
`;
